src/test/java/org/opensearch/knn/recall/RecallTestsIT.java

```diff
@@ -0,0 +1,183 @@
+import org.opensearch.knn.KNNResult;
+import org.opensearch.knn.TestUtils;
+import org.opensearch.knn.index.KNNQueryBuilder;
+import org.opensearch.knn.index.SpaceType;
+import org.opensearch.knn.plugin.script.KNNScoringUtil;
+import org.opensearch.rest.RestStatus;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import static org.opensearch.knn.index.KNNSettings.KNN_ALGO_PARAM_INDEX_THREAD_QTY;
+import static org.opensearch.knn.index.KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED;
+
+public class RecallTestsIT extends KNNRestTestCase {
+
+    private final String testFieldName = "test-field";
+    private final int dimensions = 50;
+    private final int docCount = 10000;
+    private final int queryCount = 100;
+    private final int k = 5;
+    private final double expRecallValue = 1.0;
+
+    public void testRecallL2StandardData() throws Exception {
+        String testIndexStandard = "test-index-standard";
+
+        addDocs(testIndexStandard, testFieldName, dimensions, docCount, true);
+        double recallValue = getRecallValue(testIndexStandard, testFieldName, docCount, queryCount, k, dimensions, SpaceType.L2, true);
+        assertEquals(expRecallValue, recallValue, 0.2);
+    }
+
+    public void testRecallL2RandomData() throws Exception {
+        String testIndexRandom = "test-index-random";
+
+        addDocs(testIndexRandom, testFieldName, dimensions, docCount, false);
+        double recallValue = getRecallValue(testIndexRandom, testFieldName, docCount, queryCount, k, dimensions, SpaceType.L2, false);
+        assertEquals(expRecallValue, recallValue, 0.2);
+    }
+
+    public void addDocs(String testIndex, String testField, int dimensions, int docCount, boolean isStandard) throws Exception{
+        float[][] indexVectors;
+
+        createKnnIndex(testIndex, getKNNDefaultIndexSettings(), createKnnIndexMapping(testField, dimensions));
+        updateClusterSettings(KNN_ALGO_PARAM_INDEX_THREAD_QTY, 2);
+        updateClusterSettings(KNN_MEMORY_CIRCUIT_BREAKER_ENABLED, true);
+
+        if(isStandard){
+            indexVectors = TestUtils.generateVectors(docCount, dimensions);
+        }
+        else
+            indexVectors = TestUtils.generateRandomVectors(docCount, dimensions);
+
+        bulkAddKnnDocs(testIndex, testField, indexVectors, docCount);
+    }
+
+    public void bulkAddKnnDocs(String index, String fieldName,  float[][] indexVectors, int docCount) throws IOException {
+        Request request = new Request(
+                "POST",
+                 "/_bulk"
+        );
+
+        request.addParameter("refresh", "true");
+        String body = "";
+
+        for(int i = 0; i < docCount; i++){
+          body += "{ \"index\" : { \"_index\" : \"" + index + "\", \"_id\" : \"" + String.valueOf(i+1) + "\" } }\n"
+          +"{ \"" + fieldName + "\" : "+ Arrays.toString(indexVectors[i]) +" }\n";
```

Comment:
Using a StringBuilder and append should save some resources, with current implementation there are too many string objects created during string concatenations

Reply:
updated